### PR TITLE
Separate mysql server

### LINKED ISSUE #2301: Run MySQL server out of container sqlflow:ci
In our CI job with `SQLFLOW_TEST_DB=hive`, we run HIve in a container in parallel with the SQLFlow server container

https://github.com/sql-machine-learning/sqlflow/blob/7a3bf24be17b989d43b38948531f0e40ca98bf6c/.travis.yml#L55-L59

We want to change the `SQLFLOW_TEST_DB=mysql` this way either.

After the change, we can follow the setting in sqlflow.org/desktop.

### PR DESCRIPTION
**Pull request by wangkuiyi**
Fix https://github.com/sql-machine-learning/sqlflow/issues/2301

### CODE REVIEW

**Comment by typhoonzero** [comment]
Please note that this change may break the current quick start tutorial: https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/quick_start.md

**Comment by wangkuiyi** [comment]
> Please note that this change may break the current quick start tutorial: https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/quick_start.md

@typhoonzero Thanks for the reminder. I will fix the quick start once I get this work done.

**Comment by wangkuiyi** [comment]
Take a note:

The MySQL CI job failed with

```
--- FAIL: TestEnd2EndMySQL (497.95s)
    --- FAIL: TestEnd2EndMySQL/CaseShowDatabases (0.01s)
        main_test.go:145: 
            	Error Trace:	main_test.go:145
            	            				main_test.go:691
            	Error:      	
            	Test:       	TestEnd2EndMySQL/CaseShowDatabases
            	Messages:   	string value imdb not exist
```

**Comment by wangkuiyi** [comment]
The workflow job failed

```
    --- FAIL: TestEnd2EndWorkflow/CaseWorkflowTrainAndPredictDNN (4.76s)
        main_test.go:2278: 
            	Error Trace:	main_test.go:2278
            	Error:      	Received unexpected error:
            	            	rpc error: code = Unknown desc = SQLFlow Step: [1/4] Failed, Log: http://localhost:8001/workflows/default/sqlflow-jx5qf?tab=workflow&nodeId=sqlflow-jx5qf-53446448&sidePanel=logs:sqlflow-jx5qf-53446448:main
            	            	create DB failed: failed to ping database: dial tcp 172.17.0.7:3306: connect: connection refused
            	Test:       	TestEnd2EndWorkflow/CaseWorkflowTrainAndPredictDNN
    --- SKIP: TestEnd2EndWorkflow/CaseTrainDistributedPAIArgo (0.00s)
        main_test.go:2354: Skip PAI case.
    --- FAIL: TestEnd2EndWorkflow/CaseBackticksInSQL (4.67s)
        main_test.go:2426: 
            	Error Trace:	main_test.go:2426
            	Error:      	Received unexpected error:
            	            	rpc error: code = Unknown desc = SQLFlow Step: [1/1] Failed, Log: http://localhost:8001/workflows/default/sqlflow-dn8gg?tab=workflow&nodeId=sqlflow-dn8gg-276781100&sidePanel=logs:sqlflow-dn8gg-276781100:main
            	            	create DB failed: failed to ping database: dial tcp 172.17.0.7:3306: connect: connection refused
            	Test:       	TestEnd2EndWorkflow/CaseBackticksInSQL
    --- FAIL: TestEnd2EndWorkflow/CaseWorkflowStepErrorMessage (4.70s)
        main_test.go:2235: 
            	Error Trace:	main_test.go:2235
            	Error:      	"rpc error: code = Unknown desc = SQLFlow Step: [1/1] Failed, Log: http://localhost:8001/workflows/default/sqlflow-tvkjz?tab=workflow&nodeId=sqlflow-tvkjz-69221113&sidePanel=logs:sqlflow-tvkjz-69221113:main
            	            	create DB failed: failed to ping database: dial tcp 172.17.0.7:3306: connect: connection refused" does not contain "runSQLProgram error: unsupported attribute model.no_exists_param"
            	Test:       	TestEnd2EndWorkflow/CaseWorkflowStepErrorMessage
FAIL
```

**Comment by wangkuiyi** [comment]
https://github.com/sql-machine-learning/sqlflow/pull/2306/commits/887aa915eaa417d93b8400c3d96cb2d5872e8020 aims to fix https://github.com/sql-machine-learning/sqlflow/pull/2306#issuecomment-631054994

**Comment by wangkuiyi** [comment]
The workflow CI job still fails https://travis-ci.com/github/sql-machine-learning/sqlflow/jobs/337255484#L1108-L1131

```
--- FAIL: TestEnd2EndWorkflow (15.30s)
    --- SKIP: TestEnd2EndWorkflow/CaseWorkflowTrainAndPredictDNNCustomImage (0.00s)
        main_test.go:2284: Skip PAI case.
    --- FAIL: TestEnd2EndWorkflow/CaseWorkflowTrainAndPredictDNN (4.77s)
        main_test.go:2279: 
            	Error Trace:	main_test.go:2279
            	Error:      	Received unexpected error:
            	            	rpc error: code = Unknown desc = SQLFlow Step: [1/4] Failed, Log: http://localhost:8001/workflows/default/sqlflow-dtsnq?tab=workflow&nodeId=sqlflow-dtsnq-3953869642&sidePanel=logs:sqlflow-dtsnq-3953869642:main
            	Test:       	TestEnd2EndWorkflow/CaseWorkflowTrainAndPredictDNN
    --- SKIP: TestEnd2EndWorkflow/CaseTrainDistributedPAIArgo (0.00s)
        main_test.go:2355: Skip PAI case.
    --- FAIL: TestEnd2EndWorkflow/CaseBackticksInSQL (4.70s)
        main_test.go:2427: 
            	Error Trace:	main_test.go:2427
            	Error:      	Received unexpected error:
            	            	rpc error: code = Unknown desc = SQLFlow Step: [1/1] Failed, Log: http://localhost:8001/workflows/default/sqlflow-l2kks?tab=workflow&nodeId=sqlflow-l2kks-3055913011&sidePanel=logs:sqlflow-l2kks-3055913011:main
            	Test:       	TestEnd2EndWorkflow/CaseBackticksInSQL
    --- FAIL: TestEnd2EndWorkflow/CaseWorkflowStepErrorMessage (4.71s)
        main_test.go:2236: 
            	Error Trace:	main_test.go:2236
            	Error:      	"rpc error: code = Unknown desc = SQLFlow Step: [1/1] Failed, Log: http://localhost:8001/workflows/default/sqlflow-bpnnp?tab=workflow&nodeId=sqlflow-bpnnp-3865220646&sidePanel=logs:sqlflow-bpnnp-3865220646:main
            	            	" does not contain "runSQLProgram error: unsupported attribute model.no_exists_param"
            	Test:       	TestEnd2EndWorkflow/CaseWorkflowStepErrorMessage
FAIL
FAIL	sqlflow.org/sqlflow/cmd/sqlflowserver	15.335s
```

Need help from @Yancey1989 and @typhoonzero .

It seems that the `t.Run(CaseName)` doesn't print the accurate error locate -- the line number `main_test.go:2236` is not where the error happened, but the last line of the Go function `CaseName`.  Why did we choose to run `t.Run` other than defining a test function?

**Comment by wangkuiyi** [comment]
@typhoonzero @Yancey1989 It seems that this error exposed a fact that we had been reusing `sqlflow:ci` as the step image and the MySQL server image in CI tests.  Am I right?

I noticed that in the definition of test case `CaseWorkflowTrainAndPredictDNNCustomImage`, we got the step image name from the environment variable `SQLFLOW_WORKFLOW_STEP_IMAGE`.

https://github.com/sql-machine-learning/sqlflow/blob/9f3c09661ffe0762401e1c910842b4dea9b277a1/cmd/sqlflowserver/main_test.go#L2281-L2287

This environment variable is set in `.travis.yml` as

https://github.com/sql-machine-learning/sqlflow/blob/1f85a5a67809e80bf4421987cad4dc6048840aee/.travis.yml#L93-L99

In the above `docker run ... test/workflow.sh`, we also use the environment variable `SQLFLOW_WORKFLOW_STEP_IMAGE` to specify the MySQL server image.

https://github.com/sql-machine-learning/sqlflow/blob/1f85a5a67809e80bf4421987cad4dc6048840aee/scripts/test/workflow.sh#L66-L73

File: scripts/test/workflow.sh [hunk 1]
```diff
@@ -28,88 +30,89 @@ pip -q install -r python/couler/requirements.txt
...
+POD=$(kubectl get pod -l run=mysql -o jsonpath="{.items[0].metadata.name}")
+
+TIMEOUT="true"
+for _ in {1..30}; do
+    MYSQL_POD_STATUS=$(kubectl get pod "$POD" -o jsonpath='{.status.phase}')
+    echo "${MYSQL_POD_STATUS}"
+    if [[ "${MYSQL_POD_STATUS}" == "Running" ]]; then
+        MYSQL_POD_IP=$(kubectl get pod "$POD" -o jsonpath='{.status.podIP}')
+        echo "MySQL pod IP: $MYSQL_POD_IP"
+        export SQLFLOW_TEST_DATASOURCE="mysql://root:root@tcp(${MYSQL_POD_IP}:3306)/?maxAllowedPacket=0"
```
Comment (Yancey0623) [inline scripts/test/workflow.sh:79]: From the logs of travis-CI:
``` text
create DB failed: failed to ping database: dial tcp 172.17.0.7:3306: connect: connection refused" does not contain "runSQLProgram error: unsupported attribute model.no_exists_param"
```

It seems the MySQL Pod is not ready, maybe we can add `kubectl logs ${POD}` to help debug the failed task?

Comment (Yancey0623) [reply scripts/test/workflow.sh:79]: It seems `Running` status can not mean MySQL pod is ready. I tried to use [readiness](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-readiness-probes) to wait until the population has done.

**Comment by wangkuiyi** [comment]
LGTM+++

**Review by Yancey0623 — APPROVED** [review]

**Pull request merged**